In [1]:
import pandas as pd
import numpy as np


In [2]:
ds=pd.read_csv('mushrooms.csv')
ds

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

There are 8124 rows and 23 columns in our dataset.

In [3]:
df=pd.DataFrame(data=ds)
df

class cap-shape cap-surface cap-color bruises odor gill-attachment  \
0        p         x           s         n       t    p               f   
1        e         x           s         y       t    a               f   
2        e         b           s         w       t    l               f   
3        p         x           y         w       t    p               f   
4        e         x           s         g       f    n               f   
...    ...       ...         ...       ...     ...  ...             ...   
8119     e         k           s         n       f    n               a   
8120     e         x           s         n       f    n               a   
8121     e         f           s         n       f    n               a   
8122     p         k           y         n       f    y               f   
8123     e         x           s         n       f    n               a   

     gill-spacing gill-size gill-color  ... stalk-surface-below-ring  \
0               c         n          k  ...                        s   
1               c         b          k  ...                        s   
2               c         b          n  ...                        s   
3               c         n          n  ...                        s   
4               w         b          k  ...                        s   
...           ...       ...        ...  ...                      ...   
8119            c         b          y  ...                        s   
8120            c         b          y  ...                        s   
8121            c         b          n  ...                        s   
8122            c         n          b  ...                        k   
8123            c         b          y  ...                        s   

     stalk-color-above-ring stalk-color-below-ring veil-type veil-color  \
0                         w                      w         p          w   
1                         w                      w         p          w   
2                         w                      w         p          w   
3                         w                      w         p          w   
4                         w                      w         p          w   
...                     ...                    ...       ...        ...   
8119                      o                      o         p          o   
8120                      o                      o         p          n   
8121                      o                      o         p          o   
8122                      w                      w         p          w   
8123                      o                      o         p          o   

     ring-number ring-type spore-print-color population habitat  
0              o         p                 k          s       u  
1              o         p                 n          n       g  
2              o         p                 n          n       m  
3              o         p                 k          s       u  
4              o         e                 n          a       g  
...          ...       ...               ...        ...     ...  
8119           o         p                 b          c       l  
8120           o         p                 b          v       l  
8121           o         p                 b          c       l  
8122           o         e                 w          v       l  
8123           o         p                 o          c       l  

[8124 rows x 23 columns]

By looking the dataset we can see that it's a Logistic Regression  problem.

Now we are just importing the required libraries

In [4]:
import sklearn
import matplotlib.pyplot as plt
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score
from sklearn.metrics import  confusion_matrix,classification_report
from sklearn.model_selection import train_test_split
import warnings
warnings.filterwarnings('ignore')


# Exploratory Data Analysis

In [5]:
df.columns

Index(['class', 'cap-shape', 'cap-surface', 'cap-color', 'bruises', 'odor',
       'gill-attachment', 'gill-spacing', 'gill-size', 'gill-color',
       'stalk-shape', 'stalk-root', 'stalk-surface-above-ring',
       'stalk-surface-below-ring', 'stalk-color-above-ring',
       'stalk-color-below-ring', 'veil-type', 'veil-color', 'ring-number',
       'ring-type', 'spore-print-color', 'population', 'habitat'],
      dtype='object')

In [6]:
df.shape

(8124, 23)

In [7]:
df.dtypes

class                       object
cap-shape                   object
cap-surface                 object
cap-color                   object
bruises                     object
odor                        object
gill-attachment             object
gill-spacing                object
gill-size                   object
gill-color                  object
stalk-shape                 object
stalk-root                  object
stalk-surface-above-ring    object
stalk-surface-below-ring    object
stalk-color-above-ring      object
stalk-color-below-ring      object
veil-type                   object
veil-color                  object
ring-number                 object
ring-type                   object
spore-print-color           object
population                  object
habitat                     object
dtype: object

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8124 entries, 0 to 8123
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   class                     8124 non-null   object
 1   cap-shape                 8124 non-null   object
 2   cap-surface               8124 non-null   object
 3   cap-color                 8124 non-null   object
 4   bruises                   8124 non-null   object
 5   odor                      8124 non-null   object
 6   gill-attachment           8124 non-null   object
 7   gill-spacing              8124 non-null   object
 8   gill-size                 8124 non-null   object
 9   gill-color                8124 non-null   object
 10  stalk-shape               8124 non-null   object
 11  stalk-root                8124 non-null   object
 12  stalk-surface-above-ring  8124 non-null   object
 13  stalk-surface-below-ring  8124 non-null   object
 14  stalk-color-above-ring  

In [9]:
df.isnull()

class  cap-shape  cap-surface  cap-color  bruises   odor  \
0     False      False        False      False    False  False   
1     False      False        False      False    False  False   
2     False      False        False      False    False  False   
3     False      False        False      False    False  False   
4     False      False        False      False    False  False   
...     ...        ...          ...        ...      ...    ...   
8119  False      False        False      False    False  False   
8120  False      False        False      False    False  False   
8121  False      False        False      False    False  False   
8122  False      False        False      False    False  False   
8123  False      False        False      False    False  False   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0               False         False      False       False  ...   
1               False         False      False       False  ...   
2               False         False      False       False  ...   
3               False         False      False       False  ...   
4               False         False      False       False  ...   
...               ...           ...        ...         ...  ...   
8119            False         False      False       False  ...   
8120            False         False      False       False  ...   
8121            False         False      False       False  ...   
8122            False         False      False       False  ...   
8123            False         False      False       False  ...   

      stalk-surface-below-ring  stalk-color-above-ring  \
0                        False                   False   
1                        False                   False   
2                        False                   False   
3                        False                   False   
4                        False                   False   
...                        ...                     ...   
8119                     False                   False   
8120                     False                   False   
8121                     False                   False   
8122                     False                   False   
8123                     False                   False   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
0                      False      False       False        False      False   
1                      False      False       False        False      False   
2                      False      False       False        False      False   
3                      False      False       False        False      False   
4                      False      False       False        False      False   
...                      ...        ...         ...          ...        ...   
8119                   False      False       False        False      False   
8120                   False      False       False        False      False   
8121                   False      False       False        False      False   
8122                   False      False       False        False      False   
8123                   False      False       False        False      False   

      spore-print-color  population  habitat  
0                 False       False    False  
1                 False       False    False  
2                 False       False    False  
3                 False       False    False  
4                 False       False    False  
...                 ...         ...      ...  
8119              False       False    False  
8120              False       False    False  
8121              False       False    False  
8122              False       False    False  
8123              False       False    False  

[8124 rows x 23 columns]

In [10]:
df.isnull().sum

<bound method DataFrame.sum of       class  cap-shape  cap-surface  cap-color  bruises   odor  \
0     False      False        False      False    False  False   
1     False      False        False      False    False  False   
2     False      False        False      False    False  False   
3     False      False        False      False    False  False   
4     False      False        False      False    False  False   
...     ...        ...          ...        ...      ...    ...   
8119  False      False        False      False    False  False   
8120  False      False        False      False    False  False   
8121  False      False        False      False    False  False   
8122  False      False        False      False    False  False   
8123  False      False        False      False    False  False   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0               False         False      False       False  ...   
1               False         False      F

So we can see that no null data is present in our dataset.

Here we can see that our whole dataset is present in  string format. So we have to change it in integer format.

We can done  this by applying Label Encoding on the dataset.

In [11]:
from sklearn import preprocessing

In [12]:
le=preprocessing.LabelEncoder()

In [13]:
dt=df.apply(le.fit_transform)
dt

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-below-ring  stalk-color-above-ring  \
0                            2                       7   
1                            2                       7   
2                            2                       7   
3                            2                       7   
4                            2                       7   
...                        ...                     ...   
8119                         2                       5   
8120                         2                       5   
8121                         2                       5   
8122                         1                       7   
8123                         2                       5   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
0                          7          0           2            1          4   
1                          7          0           2            1          4   
2                          7          0           2            1          4   
3                          7          0           2            1          4   
4                          7          0           2            1          0   
...                      ...        ...         ...          ...        ...   
8119                       5          0           1            1          4   
8120                       5          0           0            1          4   
8121                       5          0           1            1          4   
8122                       7          0           2            1          0   
8123                       5          0           1            1          4   

      spore-print-color  population  habitat  
0                     2           3        5  
1                     3           2        1  
2                     3           2        3  
3                     2           3        5  
4                     3           0        1  
...                 ...         ...      ...  
8119                  0           1        2  
8120                  0           4        2  
8121                  0           1        2  
8122                  7           4        2  
8123                  4           1        2  

[8124 rows x 23 columns]

In [14]:
dt['class'].unique()

array([1, 0])

In [15]:
dt['class']

0       1
1       0
2       0
3       1
4       0
       ..
8119    0
8120    0
8121    0
8122    1
8123    0
Name: class, Length: 8124, dtype: int32

Now we are having our class  and whole dataset is in  integer format.

In [16]:
mushroom_target=pd.DataFrame(dt['class'])

In [17]:
mushroom_target

class
0         1
1         0
2         0
3         1
4         0
...     ...
8119      0
8120      0
8121      0
8122      1
8123      0

[8124 rows x 1 columns]

In [18]:
y=mushroom_target
y

class
0         1
1         0
2         0
3         1
4         0
...     ...
8119      0
8120      0
8121      0
8122      1
8123      0

[8124 rows x 1 columns]

Till here y is created

In [19]:
mushroom_data=pd.DataFrame(dt)

In [20]:
x=mushroom_data
x

class  cap-shape  cap-surface  cap-color  bruises  odor  \
0         1          5            2          4        1     6   
1         0          5            2          9        1     0   
2         0          0            2          8        1     3   
3         1          5            3          8        1     6   
4         0          5            2          3        0     5   
...     ...        ...          ...        ...      ...   ...   
8119      0          3            2          4        0     5   
8120      0          5            2          4        0     5   
8121      0          2            2          4        0     5   
8122      1          3            3          4        0     8   
8123      0          5            2          4        0     5   

      gill-attachment  gill-spacing  gill-size  gill-color  ...  \
0                   1             0          1           4  ...   
1                   1             0          0           4  ...   
2                   1             0          0           5  ...   
3                   1             0          1           5  ...   
4                   1             1          0           4  ...   
...               ...           ...        ...         ...  ...   
8119                0             0          0          11  ...   
8120                0             0          0          11  ...   
8121                0             0          0           5  ...   
8122                1             0          1           0  ...   
8123                0             0          0          11  ...   

      stalk-surface-below-ring  stalk-color-above-ring  \
0                            2                       7   
1                            2                       7   
2                            2                       7   
3                            2                       7   
4                            2                       7   
...                        ...                     ...   
8119                         2                       5   
8120                         2                       5   
8121                         2                       5   
8122                         1                       7   
8123                         2                       5   

      stalk-color-below-ring  veil-type  veil-color  ring-number  ring-type  \
0                          7          0           2            1          4   
1                          7          0           2            1          4   
2                          7          0           2            1          4   
3                          7          0           2            1          4   
4                          7          0           2            1          0   
...                      ...        ...         ...          ...        ...   
8119                       5          0           1            1          4   
8120                       5          0           0            1          4   
8121                       5          0           1            1          4   
8122                       7          0           2            1          0   
8123                       5          0           1            1          4   

      spore-print-color  population  habitat  
0                     2           3        5  
1                     3           2        1  
2                     3           2        3  
3                     2           3        5  
4                     3           0        1  
...                 ...         ...      ...  
8119                  0           1        2  
8120                  0           4        2  
8121                  0           1        2  
8122                  7           4        2  
8123                  4           1        2  

[8124 rows x 23 columns]

Now we have created both x and y for train_test_split

In [21]:
x.shape

(8124, 23)

In [22]:
y.shape

(8124, 1)

In [33]:
train_x,test_x,train_y,test_y=train_test_split(x,y,test_size=.25,random_state=53)

In [24]:
train_x.shape

(6093, 23)

In [25]:
test_x.shape

(2031, 23)

In [26]:
train_y.shape

(6093, 1)

In [27]:
test_y.shape

(2031, 1)

Now we use different models for prediction.

In [30]:
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier



To save time we use a for loop to run all the models.

In [34]:
model=[DecisionTreeClassifier(),SVC(),KNeighborsClassifier(),GaussianNB(),LogisticRegression()]

for m in model:
    m.fit(train_x,train_y)
    m.score(train_x,train_y)
    predm=m.predict(test_x)
    print('Accuracy score of ',m,'is:')
    print(accuracy_score(test_y,predm))
    print(confusion_matrix(test_y,predm))
    print(classification_report(test_y,predm))
    print('\n')
          


Accuracy score of  DecisionTreeClassifier() is:
1.0
[[1079    0]
 [   0  952]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1079
           1       1.00      1.00      1.00       952

    accuracy                           1.00      2031
   macro avg       1.00      1.00      1.00      2031
weighted avg       1.00      1.00      1.00      2031



Accuracy score of  SVC() is:
1.0
[[1079    0]
 [   0  952]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1079
           1       1.00      1.00      1.00       952

    accuracy                           1.00      2031
   macro avg       1.00      1.00      1.00      2031
weighted avg       1.00      1.00      1.00      2031



Accuracy score of  KNeighborsClassifier() is:
0.9985228951255539
[[1076    3]
 [   0  952]]
              precision    recall  f1-score   support

           0       1.00      1.00      1.00      1079
  

WE can see that all models except KNeighborsClassifier() has 100 % accuracy .
So now we will check for cross validation and get our best model.

# Cross Validation

In [39]:
from sklearn.model_selection import cross_val_score
gnbscores=cross_val_score(GaussianNB(),x,y,cv=5)
print(gnbscores)
print(gnbscores.mean(),gnbscores.std())

[1. 1. 1. 1. 1.]
1.0 0.0


In [40]:
svcscores =cross_val_score(SVC(),x,y,cv=5)
print(svcscores)
print(svcscores.mean(),svcscores.std())

[0.95076923 1.         1.         1.         0.68349754]
0.9268533535430088 0.12316275274521919


In [41]:
dtcscores =cross_val_score(DecisionTreeClassifier(),x,y,cv=5)
print(dtcscores)
print(dtcscores.mean(),dtcscores.std())

[1. 1. 1. 1. 1.]
1.0 0.0


In [43]:
knnscores=cross_val_score(KNeighborsClassifier(),x,y,cv=5)
print(knnscores)
print(knnscores.mean(),knnscores.std())

[0.85415385 0.99815385 0.94030769 1.         0.6453202 ]
0.8875871163319438 0.13225139774323946


In [44]:
lgscores=cross_val_score(LogisticRegression(),x,y,cv=5)
print(lgscores)
print(lgscores.mean(),lgscores.std())

[1.         1.         1.         1.         0.97475369]
0.9949507389162562 0.010098522167487679


By Looking  the results from cross validation we can see that we are getting 100% accuracy from GaussianNB and DecisionTreeClassifier.


So I am choosing DecisionTreeClassifier as my best model.

# Saving the Best Model

In [ ]:
import pickle

filename='pickleDecisionTreeClassifierfile.pkl'
pickle.dump(DecisionTreeClassifier,open(filename, 'wb'))

loaded_model = pickle.load(open(filename,'rb'))
loaded_model.predict(test_x)